src/ProjectOrigin.WalletSystem.IntegrationTests/Repositories/CertificateRepositoryTests.cs

```diff
@@ -173,6 +174,71 @@ public async Task GetAllOwnedCertificates()
         );
     }
 
+    [Fact]
+    public async Task GetAllOwnedCertificates_WalletAttributes()
+    {
+        // Arrange
+        var endpointPosition = 1;
+        var registry = _fixture.Create<string>();
+        var owner = _fixture.Create<string>();
+        var wallet = await CreateWallet(owner);
+        var endpoint = await CreateWalletEndpoint(wallet);
+
+        var certificateId = Guid.NewGuid();
+        WalletAttribute walletAttribute = new WalletAttribute
+        {
+            Key = "AssetId",
+            Value = "571234567890123456",
+            CertificateId = certificateId,
+            RegistryName = registry,
+            Salt = Encoding.UTF8.GetBytes(Guid.NewGuid().ToString())
+        };
+
+        var attributes = new List<CertificateAttribute>
+        {
+            new(){ Key="AssetId", Value=walletAttribute.GetHashedValue(), Type=CertificateAttributeType.Hashed},
+            new(){ Key="TechCode", Value="T070000", Type=CertificateAttributeType.ClearText},
+            new(){ Key="FuelCode", Value="F00000000", Type=CertificateAttributeType.ClearText},
+        };
+        var certificate = new Certificate
+        {
+            Id = certificateId,
+            RegistryName = registry,
+            StartDate = DateTimeOffset.Now.ToUtcTime(),
+            EndDate = DateTimeOffset.Now.AddDays(1).ToUtcTime(),
+            GridArea = "DK1",
+            CertificateType = GranularCertificateType.Production,
+            Attributes = attributes
+        };
+
+        var slice = new WalletSlice
+        {
+            Id = Guid.NewGuid(),
+            WalletEndpointId = endpoint.Id,
+            WalletEndpointPosition = endpointPosition,
+            RegistryName = registry,
+            CertificateId = certificate.Id,
+            Quantity = _fixture.Create<int>(),
+            RandomR = _fixture.Create<byte[]>(),
+            State = WalletSliceState.Available
+        };
+
+        await _repository.InsertCertificate(certificate);
+        await _repository.InsertWalletAttribute(wallet.Id, walletAttribute);
+        await _repository.InsertWalletSlice(slice);
+
+        // Act
+        var certificates = await _repository.GetAllOwnedCertificates(owner);
+
+        // Assert
+        certificates.Should().HaveCount(1).And.Satisfy(
+            c => c.Id == certificate.Id
+                 && c.Slices.Sum(x => x.Quantity) == slice.Quantity
+                 && c.Attributes.Count == 3
+                 && c.Attributes.Any(x => x.Key == walletAttribute.Key && x.Value == walletAttribute.Value)
```

Comment:
There should only be one match
```suggestion
                 && c.Attributes.Single(x => x.Key == walletAttribute.Key && x.Value == walletAttribute.Value)
```

Reply:
done